src/resource.rs

```diff
@@ -344,18 +348,24 @@ impl Resource {
     }
 
     /// Add an exported type to the resource.
-    pub fn add_type(&mut self, ty: TypeDef) {
+    pub fn add_type(&mut self, ty: TypeDef) -> Result<(), String> {
+        if ty.resource != ResourceId::default() {
```

Comment:
This never updates the empty `ty.resource`?

Maybe we should use an `Option<SmolStr>` to flag when the definition hasn't been attached to a resource, and overwrite it before inserting it to `self.types`? (and do the same for OpDef)

I just checked, and `SmolStr` works well with NPO (it takes 24 bytes both with and without the `Option`).

Reply:
sorry yeah not updating the resource was an oversight, took your suggestion